File: cpp/gql/src/libgql/src/libgql/parsers/schema/nodes/input.cpp
#include "./input.hpp"

#include <map>
#include <memory>
#include <ranges>
#include <string>
#include <utility>

#include "../../file/server/ast.hpp"
#include "../../file/shared/ast.hpp"
#include "../server_ast.hpp"
#include "../shared_ast.hpp"
#include "../type_registry.hpp"
#include "./input_type_spec.hpp"

using namespace gql::parsers::file;

namespace gql::parsers::schema::nodes {

std::shared_ptr<ast::InputType> parseInput(
    const server::ast::InputObjectDefinitionNode &node,
    const TypeRegistry &registry) {
    const auto &obj = registry.inputs.at(node.name.name);
    obj->fields =
        node.fields |
        std::views::transform(
            [&registry](const shared::ast::InputFieldDefinitionNode &defNode)
                -> std::pair<std::string,
                             ast::FieldDefinition<ast::InputFieldSpec>> {
                const auto &[typeSpec, nullable] =
                    nodes::parseInputTypeSpec(defNode, registry);
                return { defNode.name.name,
                         { .name = defNode.name.name,
                           .spec = typeSpec,
                           .nullable = nullable } };
            }) |
        std::ranges::to<std::map>();
    return obj;
};

};  // namespace gql::parsers::schema::nodes

File: rust/gql/libgql/src/parsers/schema/server/input.rs
use std::{cell::RefCell, rc::Rc};

use super::errors;
use crate::parsers::{
    file::server::ast::InputObjectDefinitionNode,
    schema::{shared, type_registry::TypeRegistry},
};

pub fn parse_definition(
    input: &InputObjectDefinitionNode,
    registry: &mut TypeRegistry,
) -> Result<Rc<RefCell<shared::ast::InputType>>, errors::Error> {
    let obj_rc = registry.inputs.get(&input.name.name).unwrap();
    let mut obj = obj_rc.borrow_mut();
    obj.fields =
        shared::input::parse_field_definitions(&input.fields, registry)?;
    return Ok(obj_rc.clone());
}
